### PR TITLE
Able to print gradients in event_handler

### PR DESCRIPTION
**Pull request by typhoonzero**
Able to print gradients in event handle using v2 API to train.

Related: https://github.com/PaddlePaddle/Paddle/issues/3040
Fixes: https://github.com/PaddlePaddle/Paddle/issues/3211

```python
def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            if event.batch_id % 10 == 0:
                for p in parameters:
                    print "parameters:", parameters.get(p)
                    grad = parameters.get_grad(p)
                    print "gradients:", grad
                    print "gradients max abs:" max(grad.min(), grad.max(), key=abs)
            if event.batch_id % 100 == 0:
                print "Pass %d, Batch %d, Cost %f" % (
                    event.pass_id, event.batch_id, event.cost)
```

### CODE REVIEW

File: python/paddle/v2/trainer.py [hunk 1]
```diff
@@ -161,14 +161,14 @@ def train(self, reader, num_passes=1, event_handler=None, feeding=None):
                     self.__parameter_updater__.update(each_param)
                 cost_sum = out_args.sum()
                 cost = cost_sum / len(data_batch)
-                self.__parameter_updater__.finishBatch(cost)
```
Comment (Yancey1989) [inline python/paddle/v2/trainer.py]: Just a little confusing, why we move the `batch_evaluator` to the `event_handler`?

File: python/paddle/v2/parameters.py [hunk 2]
```diff
@@ -138,14 +129,27 @@ def __getitem__(self, key):
...
-                val = param.getBuf(api.PARAMETER_VALUE)
+                val = param.getBuf(param_type)
                 assert isinstance(val, api.Vector)
                 val = val.copyToNumpyArray()
                 return val
                 # else continue
 
             raise RuntimeError("Unexpected branch")
 
+    def __getitem__(self, key):
```
Comment (jacquesqiao) [inline python/paddle/v2/parameters.py]: rename to get_param?

Comment (typhoonzero) [inline python/paddle/v2/parameters.py]: `__getitem__` will called when doing `param[k]`, it's an operator reload. Need to keep this the same as before.

File: python/paddle/v2/trainer.py (hunk 1 above)
Comment (typhoonzero) [inline python/paddle/v2/trainer.py]: Need to call `event_handler` before finishBatch operations so we can get inner status before clearing them.

Comment (Yancey1989) [inline python/paddle/v2/trainer.py]: Got it, thanks!

**Comment by lcy-seso** [comment]
even though it copies the parameter from the c++ side, very helpful for debugging the training process and tuning the model,  thank you very much.

**Pull request merged**